### PR TITLE
[Refactor] API Re-Design

### LINKED ISSUE #344: Metric API re-design
## 🚀 Feature

Re-design the internal API for Base Metric class to return batch states from the update function.

**The following proposal is a BC !**

### Limitations:
Currently, the Metric API is confusing for several reasons:
* The `update` function actually performs a `compute` if `compute_on_step=True`
* The `update` function actually perform 2 updates, which is an expensive operation for some metrics, FID for example
* Users don't have a clear API to perform computation
* The Metric internals is tailored to Lightning. IMO, TM should define its own API to better faciliate metrics computation and Lightning should adapt to it.

### Proposal:
* The update function doesn't return anything
* There is 2 functions for reduction, `compute` and `compute_step`
* The update function is responsible to return a dictionary containing the batch states

Here is the internal re-design of Metric 

```py
class Metric()

    def _wrap_update()
        batch_states = update(...)
        self.add_to_rank_accumulated_states(batch_states) # uses reduction function
        self.batch_states = batch_states

class MyMetric(Metric):
    def update(self) -> Dict[str, Tensor]:
        ...
        return {"state_1": state_1, ...}

metric = MyMetric()

metric(...) # compute batch states, add them to accumulated states using reduction functions
compute() # accumulated_states compute on all ranks
compute_on_step() # batch_states compute on all ranks
compute(sync_dist=False) # accumulated_states compute per rank
compute_on_step(sync_dist=False) #  batch_states compute per rank 
```

```py
class Accuracy

    def __init__(self):
        self.add_state("correct", torch.tensor(0.), sync_dist_fn=torch.sum)
        self.add_state("total", torch.tensor(0.), sync_dist_fn=torch.sum)

    def update(self, preds, targets):
        return {"total": preds.shape[0], "correct": (preds == targets).sum()}

    def compute(self):
        return self.correct / self.total


metric = Accuracy()

None = metric.update([0, 1], [0, 0])

0.5 = metric([0, 1], [0, 0], sync_dist=True) # compute batch states and cache it, add batch states to accumulated states
1 = metric([0, 0], [0, 0], sync_dist=True)

0.75 = metric.compute()
1. = metric.compute(accumulated=False)
```

```
# accumulated=True means computing accuracy on 3 batches
# accumulated=False means computing accuracy on latest batch
acc()
acc()
acc()
```

### Additional context

sync = store cache +  all_gather + reduction
unsync = restore cache

### PR DESCRIPTION
**Pull request by tchaton**
# Before submitting

This PR introduce a new design API for TorchMetrics.

TODO: Add details.

Fixes #344

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?

Fixes # (issue).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Breaking Changes !

TorchMetrics forward function isn't storing the cache states anymore.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@tchaton @SkafteNicki how are we doing here? :rabbit:

**Comment by Borda** [comment]
@tchaton @SkafteNicki @maximsch2 any progress here?
